openmessaging-api/src/main/java/io/openmessaging/api/Client.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2017 OpenMessaging
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.openmessaging.api;
+
+import io.openmessaging.api.consumer.Consumer;
+import io.openmessaging.api.consumer.ConsumerConfiguration;
+import io.openmessaging.api.producer.Producer;
+import io.openmessaging.api.producer.ProducerConfiguration;
+import io.openmessaging.api.reader.Reader;
+import io.openmessaging.api.reader.ReaderConfiguration;
+import io.openmessaging.exception.OMSException;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An OMS client that talks to OMS messaging services.
+ */
+public interface Client extends AutoCloseable {
```

Comment:
`AutoCloseable` and `CompletableFuture` are specific APIs in Java platform, while we want to use very simple and common APIs  to describe our specification, to make the multilingual APIs looks consistent as much as possible~

Reply:
And we already defined the Future and Promise interfaces, please refer to [Future](http://openmessaging.cloud/openmessaging-java/io/openmessaging/Future.html).

Reply:
In my opinion the important part is that the same concepts (Client, Producer, Consumer, ...) are exposed in the API for all the supported languages. 

Every language has its own idiomatic way to do implement the concepts and we should follow that when adapting the the API concepts to that language. 

For example, in Java the native way is to use `CompletableFuture`. This is what Java programmers are used to. This is what other libraries are expecting or integrating with. 

When defining the API for Python, we then need to make sure the API is Python-like and friendly to Python programmer. For example, in Python the common way to do async APIs is to use callback functions instead of futures. We should follow that instead of pushing a Java like API to Python programmers.

Reply:
> And we already defined the Future and Promise interfaces, please refer to Future.

I most strongly recommend against inventing new `Future` and `Promise` classes. Java already provide a much better abstraction that is natively supported by many libraries and frameworks. 

By using the standard `CompletableFuture`, a Java developer can chain multiple asynchrous operations in a single complex future. If we were to do a non-standard Future, that would be impossible. 

```java
return source.getMessage()
                  .thenCombine(msg -> producer.sendAsync(msg))
                  .thenCombine(msgId -> database.put(msg.getKey(), "..something"));
```
The above is a fictitious example, but it serve to illustrate that, if we play in the idiomatic Java way, the API will fill both natural to Java programmer as well it will integrate with all the other libraries.

Reply:
We can discuss the native and common API which is better for spec.

BTW, `CompletableFuture` is imported in JDK 1.8, which is not friendly for SDK, many legacy applications even use JDK 1.6~

Reply:
@zhouxinyu : I agreed with Matteo. we should not reinvent `Future` in java api. we should use CompletableFuture to chain the futures in the asynchronous applications. If the goal is to make consistent, why not adding similar CompletableFuture in other languages, rather than giving up a good standard library in java. As what Matteo said, the most important thing is to make sure the model and api semantic are consistent between languages, not classes.

Reply:
@zhouxinyu checkout my last comment. oracle already stopped public updates on Java 7 since 2015. There is no point to support java 7 or even earlier version. I would suggest supporting 8 and forward, and leveraging all the lambda functionalities there.

Reply:
@merlimat @sijie Client concept is not suitable for standard, It is a coarse-grained implementation model. We must remember what we are doing for these days, we hope to create standard, user-friendly API standard. we must jump out from our implementation. 


@zhouxinyu I also agree we'd not better reinvent another similar Naming concept, such as future and promise. But, if we want to strengthen the JDK implementation, we could create similar but not ambiguous concept.


 
AutoCloseable is created since JDK 7, and CompletableFuture is imported from JDK 8. IMO, It comes up another argument - backward compatible. I know oracle recommend JDK 9 in its JDK download page for long times. and JDK 7 is not even supported when we use 'brew cast 'in Macbook. But, many cloud organization, such as Ali-Cloud, Amazon(refer here, https://repo1.maven.org/maven2/com/amazonaws/aws-java-sdk-pom/1.11.228/aws-java-sdk-pom-1.11.228.pom) are all supporting JDK 6 in its client SDK. For this reason, I hope we could not import high version syntax feature into our standard.  But, if you want to support more awesome coding experience for developer, we could discuss how to reach this point :-)

Reply:
@zhouxinyu Java7 was EOLd on late 2014. Large majority of companies have switched to Java8 long time ago, since there no more security updates on Java7 since then.

Reply:
@vongosling 
```
Client concept is not suitable for standard, It is a coarse-grained implementation model. 
```
`Client` is the equivalent of [MessagingAccessPoint](https://github.com/openmessaging/openmessaging-java/blob/master/openmessaging-api/src/main/java/io/openmessaging/MessagingAccessPoint.java) here. I don't understand why it is not suitable for standard...

Reply:
Calling `MessagingAccessPoint` is a bit confusing in the sense that it seems to be a "server" component rather than part of the client API and library.

Reply:
@vongosling For what is worth, AWS Java SDK is based on Java8 : https://aws.amazon.com/sdk-for-java/

> We’re pleased to announce the Developer Preview of the AWS SDK for Java 2.0. The 2.0 
> version of the SDK is a major rewrite of the 1.11.x code base. It’s built on top of Java 8 and 
> adds several, frequently requested features, like support for non-blocking I/O and
> ....